lib/ssh/src/ssh_auth.erl

```diff
@@ -140,47 +141,67 @@ keyboard_interactive_msg([#ssh{user = User,
 get_public_key(SigAlg, #ssh{opts = Opts}) ->
     KeyAlg = key_alg(SigAlg),
     case ssh_transport:call_KeyCb(user_key, [KeyAlg], Opts) of
+        {ok, {ssh2_pubkey, PubKeyBlob}} ->
+            {ok, {ssh2_pubkey, PubKeyBlob}};
+
         {ok, PrivKey} ->
             try
                 %% Check the key - the KeyCb may be a buggy plugin
                 true = ssh_transport:valid_key_sha_alg(private, PrivKey, KeyAlg),
                 Key = ssh_transport:extract_public_key(PrivKey),
                 ssh_message:ssh2_pubkey_encode(Key)
             of
-                PubKeyBlob -> {ok,{PrivKey,PubKeyBlob}}
+                PubKeyBlob -> {ok, {PrivKey, PubKeyBlob}}
             catch
                 _:_ -> 
                     not_ok
             end;
 
-	_Error ->
-	    not_ok
+        _Error ->
+            not_ok
     end.
 
 
 publickey_msg([SigAlg, #ssh{user = User,
-		       session_id = SessionId,
-		       service = Service} = Ssh]) ->
+                            session_id = SessionId,
+                            service = Service,
+                            opts = Opts} = Ssh]) ->
     case get_public_key(SigAlg, Ssh) of
-	{ok, {PrivKey,PubKeyBlob}} ->
+        {ok, {_, PubKeyBlob} = Key} ->
             SigAlgStr = atom_to_list(SigAlg),
-            SigData = build_sig_data(SessionId, User, Service,
-                                     PubKeyBlob, SigAlgStr),
-            Hash = ssh_transport:sha(SigAlg),
-            Sig = ssh_transport:sign(SigData, Hash, PrivKey),
+            SigData = build_sig_data(SessionId, User, Service, PubKeyBlob, SigAlgStr),
+
+            Sig = case Key of
+                {ssh2_pubkey, PubKeyBlob} ->
+                  {ssh_agent, KeyCbOpts} = ?GET_OPT(key_cb, Opts),
+                  % OpenSSH does not seem to care when these flags are set for
+                  % signature algorithms other than RSA, so we always send them.
+                  SignFlags = ?SSH_AGENT_RSA_SHA2_256 bor ?SSH_AGENT_RSA_SHA2_512,
+                  SignRequest = #ssh_agent_sign_request{key_blob = PubKeyBlob, data = SigData, flags = SignFlags},
+                  SignResponse = ssh_agent:send(SignRequest, KeyCbOpts),
+                  #ssh_agent_sign_response{signature = #ssh_agent_signature{blob = Blob}} = SignResponse,
```

Comment:
I don't like the idea that ssh_auth has knowledge about agents and also is bound to a certain name of the callback file.
What about a new optional callback function in the ssh_client_key_api so the call would be roughly:
```suggestion
                  Blob = ssh_transport:call_KeyCb(sign, [SigData, ....], Opts);
```

Reply:
💯 👍  We actually debated a solution pretty exactly like what you suggest. Sounds perfect.

Would you move the current signing logic (i.e. the second branch in this `case`) to a `ssh_file:sign` or should it stay where it is?

Reply:
Good question - I have to think about it.  Keep it in ssh_auth.erl, I can move it later.